### PR TITLE
flarum: add fof extensions

### PR DESCRIPTION
**Pull request by fsagbuya**
## Description of changes

Add some useful extensions from https://github.com/FriendsOfFlarum:
- [fof/upload](https://github.com/FriendsOfFlarum/upload)
- [fof/polls](https://github.com/FriendsOfFlarum/polls)
- [fof/subscribed](https://github.com/FriendsOfFlarum/subscribed)

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/by-name/fl/flarum/package.nix [hunk 1]
```diff
@@ -14,15 +16,27 @@ php.buildComposerProject (finalAttrs: {
     hash = "sha256-kigUZpiHTM24XSz33VQYdeulG1YI5s/M02V7xue72VM=";
   };
 
+  patches = [
+    # Add useful extensions from https://github.com/FriendsOfFlarum
+    # Extensions included: fof/upload, fof/polls, fof/subscribed
+    (fetchpatch {
+      url = "https://github.com/flarum/flarum/commit/aa4c8148f5614f437c9e7be4a5364e9d3303606f.patch";
```
Comment (drupol) [inline pkgs/by-name/fl/flarum/package.nix:23]: Where is the PR corresponding to this commit?

Comment (drupol) [reply pkgs/by-name/fl/flarum/package.nix:23]: This commit comes from your fork at https://github.com/fsagbuya/flarum/commit/aa4c8148f5614f437c9e7be4a5364e9d3303606f

Therefore, I would definitely favour a commit coming straight from the official project or switch back to a custom patch.

**Comment by fsagbuya** [comment]
Hi @drupol. If everything looks good, I'd like to request your approval to merge. Planning to use `buildComposerProject2` in the future, potentially when the flarum package receives a version update. Thank you for your review.

**Comment by jtojnar** [comment]
:-1: from me. This will make setups that need these three extensions easier at the cost of making it harder if you need any other extension (which I think most flarum installations will need eventually). To install other extensions, you would need to create a similar patch against flarum/flarum with this patch applied, rather than just against flarum/flarum.

The issue with installation of extensions is also why I did not contribute my Flarum package to Nixpkgs in the first place.

**Comment by fsagbuya** [comment]
@jtojnar I agree with your points about the potential drawbacks of including specific extensions by default. You've raised valid concerns about maintainability and flexibility for different Flarum installations.

Given these considerations, I'll keep this change local for now and not pursue upstreaming it. I hope we can find a more flexible solution for managing Flarum extensions in `nixpkgs` in the future, perhaps through improvements to `buildComposerProject2` or a new approach.

I appreciate your inputs. Will close this PR now.

**Comment by fsagbuya** [comment]
> It might be possible to find a sufficiently wide set of popular extensions that could be installed by default to satisfy most use cases, which would outweigh that cost

Could we create a separate package of Flarum with those extensions? This way, we'd have a basic/minimal Flarum package and another one with extensions. What do you think? @drupol @jtojnar

**Comment by jtojnar** [comment]
Unfortunately, since PHP has global name scope, we would need to ensure that all extensions use the same versions of dependencies as flarum and other extensions, which is generally a pain (see `node-modules.nix`). Or we could automatically rename all extensions dependencies (e.g. PHPStan does this), which would lead to duplication and possibly interoperability issues. 

So, in theory, it is solvable but it would be a lot of continuous work. Having each user maintain their local lockfile is a bit more work for users but it feels much more tractable.

**Comment by yoctocell** [comment]
Not directly related to the PR, but I couldn't find any info about this.  How would one install Flarum extensions when using the Nixos module?  Doing `composer require` doesn't work because it tries to write to the Nix store.

**Pull request closed without merging**